feat: Thread names. Fix warnings. Handle thread panic. Debug config. UML

ltfschoen opened the pull request.

jamesray1 commented on README.md:
````diff
@@ -67,6 +67,11 @@ For further information please refer to the [Introduction and onboarding process
 
+  * Troubleshooting:
+    ```bash
+    RUST_TEST_THREADS=5 cargo test --test cli_test && cargo test
````
I don't think we should add this to the readme, and it's too early to tell whether we should add this to the wiki. So this causes the 4 tests in cli_test to pass, but according to https://doc.rust-lang.org/cargo/guide/tests.html `cargo test` should run all the tests in each file in `src/` and any tests in `tests/`. So it's odd that the tests in `tests/` don't run with `cargo test`. We should find out how to get all the tests to run with `cargo test`.

ltfschoen replied:
Agreed, I'll remove this

jamesray1 commented on src/cli/config.rs:
```diff
@@ -5,6 +5,16 @@ pub enum Mode {
+            Mode::Proposer => "proposer".to_string(),
+            Mode::Collator => "collator".to_string(),
+            Mode::Both => "both".to_string()
+        }
```
Yeah this doesn't need to be exhaustive since this is filtered from args.rs.

ltfschoen replied:
This is to generate the name of the thread as a string. I thought this would be the best abstraction for it instead of hard-coding the values in lib.rs.
I have to include the third option `Mode::Both => "both".to_string()`, otherwise it gives an error as shown below:
![screen shot 2018-04-09 at 6 07 28 pm](https://user-images.githubusercontent.com/6226175/38486457-7623d7c0-3c20-11e8-8732-3192821a6ed4.png)

ltfschoen replied:
I've changed it to the following so that the third option catches "Both" or any other options we may add until we explicitly include them:
```
Mode::Proposer => "proposer".to_string(),
Mode::Collator => "collator".to_string(),
_ => "both".to_string()
```

ChosunOne replied:
I think it is good to catch invalid inputs though, otherwise it might do something the user didn't expect.

jamesray1 replied:
How about:
```
Mode::Proposer => "proposer".to_string(),
Mode::Collator => "collator".to_string(),
Mode::Both => "both".to_string()
_ = > "Invalid input, please try `-mode p`, `-mode proposer`, `-mode c`, `-mode collator`, `-mode b` or `-mode both`."

ltfschoen replied:
Enum values can only be one of the Variants that are defined in the Enum. If a developer were to add another Variant to `Mode` (i.e. `SomeOtherVariant`) and they didn't add it as a possible match in this Embedded Method that was created using `impl Mode`, then they wouldn't even be able to compile the code because `cargo build` would complain `error[E0004]: non-exhaustive patterns: SomeOtherVariant not covered`. 
If on the other hand they hadn't added `SomeOtherVariant` Variant to the `Mode` Enum and tried calling the Embedded Method with `Mode::SomeOtherVariant.value();` then the compiler would complain with `variant not found in cli::config::Mode`.
So whilst I don't think adding that extra line `_ = > "Invalid input, please try -mode p, -mode proposer, -mode c, -mode collator, -mode b or -mode both.".to_string()` would cause any harm... I'm not sure how to write a unit test to trigger it... I think the compiler will protect us, but I don't know the compiler well enough and it's always ignoring my calls ;-)

jamesray1 commented on src/cli/config.rs:
```diff
@@ -14,4 +24,18 @@ impl Config {
+    use super::*;
+
+    #[test]
+    fn it_sets_thread_name_for_config_arguments() {
```
I think we should have the syntax in the imperative form: `set_thread_name_for_config_arguments`.

ltfschoen replied:
Ok i'll fix that

jamesray1 replied:
@ChosunOne could you do the same please?

jamesray1 commented on README.md:
````diff
@@ -67,8 +67,22 @@ For further information please refer to the [Introduction and onboarding process
+    open -a "Google Chrome" ./diagrams/ml.svg
+    ```
+
+![](https://github.com/Drops-of-Diamond/diamond_drops/diagrams/ml.svg)
````
The diagram isn't showing.

jamesray1 replied:
Also you could use a relative link. Note that this diagram will get complicated over time, so we should probably move it to the wiki.

jamesray1 replied:
I think this doesn't work because it is an svg file. One thing to try is to automatically convert it to a [supported file type](https://blog.github.com/2015-09-25-attach-files-to-comments/) e.g. PNG.

ltfschoen replied:
It'll work with svg's. I just tested it in my fork as shown in the image below. Since we're only going to care about actually seeing it in the Readme when it's in the "master" branch (where the Readme appears), the link will be https://raw.githubusercontent.com/Drops-of-Diamond/diamond_drops/master/diagrams/ml.svg?sanitize=true.
If anyone wants to see it otherwise, they just need to run `cargo build` and then open the /diagrams/ml.svg in their browser to view the latest.

![https://raw.githubusercontent.com/ltfschoen/Diamond-drops/collator/diagrams/ml.svg?sanitize=true](https://raw.githubusercontent.com/ltfschoen/Diamond-drops/collator/diagrams/ml.svg?sanitize=true)

jamesray1 commented on src/lib.rs:
```diff
@@ -12,7 +14,7 @@ pub mod message;
 use std::sync::mpsc;
 
-pub fn run(config: cli::config::Config) {
+pub fn run(config: cli::config::Config) -> () {
```
The return should be a type, e.g. None.

jamesray1 commented on src/lib.rs:
```diff
@@ -39,35 +41,57 @@ pub fn run(config: cli::config::Config) {
+                .spawn(move || {
+                    proposer.run();
+                })
+                .expect("Failed to spawn a proposer thread")
```
When this function returns a result, we should change this.

jamesray1 replied:
Then again, maybe not, since this is just a test.

ltfschoen replied:
We can also make it panic by modifying proposer.rs file to be:
```
pub fn run(&self) { panic!("Make this test fail"); }
```
Then if you run `cargo run -- -mode p`, it'll show the thread name that we defined which paniced:
![screen shot 2018-04-09 at 6 13 53 pm](https://user-images.githubusercontent.com/6226175/38486751-4a81f038-3c21-11e8-871a-c1277009d72e.png)

If you want to check that both the Collator and the Proposer threads are both running, you can just add `panic!` with a different message in the `run` function of both proposer.rs and collator.rs, and then run it in "Both" mode `cargo run -- -mode b`, and it'll show the following:
![screen shot 2018-04-09 at 6 18 17 pm](https://user-images.githubusercontent.com/6226175/38486947-e33f0162-3c21-11e8-8e2a-047ce4aaf530.png)

jamesray1 replied:
I think it's best to avoid panics when it's not in tests, and consider alternatives. But will alternatives be suitable in this case?

https://doc.rust-lang.org/stable/book/second-edition/ch11-01-writing-tests.html

https://doc.rust-lang.org/stable/book/second-edition/ch09-02-recoverable-errors-with-result.html

ltfschoen replied:
I'm wasn't suggesting to leave a `panic!` in the `run` function, I was only explaining what would happen at the moment if an error occurred in one of the threads... i was just temporarily using `panic!` in `run` it to confirm that the thread names were being associated with the thread they were assigned to

jamesray1 replied:
OK, fair enough.

ltfschoen commented:
@jamesray1 I've pushed commit https://github.com/Drops-of-Diamond/diamond_drops/pull/15/commits/a34f515bf5fb61082a336a8a635cf6c3ec677c58, which addresses all your PR review comments

jamesray1 commented on README.md:
````diff
@@ -69,6 +69,15 @@ For further information please refer to the [Introduction and onboarding process
+    open -a "Google Chrome" ./diagrams/ml.svg
+    ```
+![](
+https://raw.githubusercontent.com/Drops-of-Diamond/diamond_drops/master/diagrams/ml.svg?sanitize=true)
````
Unfortunately there is still no image.

ltfschoen replied:
Yes, that was deliberate. As you can see in the URL that I've used, I was only asking it to render after the diagram has been merged into the 'master' branch.
I've changed now to just load the image that's been merged into my own fork

jamesray1 replied:
OK, I wasn't sure what the sanitize query meant. I was just giving an example, we can improve it if needed. But let's work on the tasks in #13.

jamesray1 replied:
OK, I wasn't sure what the sanitize query meant. I was just giving an example, we can improve it if needed. But let's work on the tasks in #13.

jamesray1 commented:
I suggest to halt work on this until a new spec is defined and work on the tasks in #13 in the meantime.

ChosunOne commented:
I'm merging this since the changes introduced don't cause any tests to fail that are related to this, I'll remove the old collator stuff if it gets included
